### PR TITLE
[CORDA-1572] Create marker file to track initial registration status.

### PR DESCRIPTION
**Pull request by florian-f**
https://r3-cev.atlassian.net/browse/CORDA-1572

### CODE REVIEW

File: node/src/main/kotlin/net/corda/node/internal/NodeStartup.kt [hunk 1]
```diff
@@ -65,13 +70,23 @@ open class NodeStartup(val args: Array<String>) {
             println("Corda will now exit...")
             return false
         }
-        val cmdlineOptions = NodeArgsParser().parseOrExit(*args)
+
+        val registrationMode = checkRegistrationMode()
+        val cmdlineOptions: CmdLineOptions
+        if (registrationMode && !args.contains("--initial-registration")) {
```
Comment (kasiastreich) [inline node/src/main/kotlin/net/corda/node/internal/NodeStartup.kt]: you can do something like `val mdLineOptions = if (blah) NodeArgsParser... else NodeArgsParser...`

File: node/src/main/kotlin/net/corda/node/internal/NodeStartup.kt [hunk 2]
```diff
@@ -65,13 +70,23 @@ open class NodeStartup(val args: Array<String>) {
...
+            // The node was started before with `--initial-registration`, but the registration has not completed.
+            // So we pretend that the node was started with `--initial-registration`
+            logger.info("Node was started before with `--initial-registration`, but the registration was not completed.\nResuming registration.")
```
Comment (kasiastreich) [inline node/src/main/kotlin/net/corda/node/internal/NodeStartup.kt]: Shouldn't it be printed to console as well?

File: node/src/main/kotlin/net/corda/node/internal/NodeStartup.kt [hunk 3]
```diff
@@ -184,6 +201,59 @@ open class NodeStartup(val args: Array<String>) {
...
+        // We do this here to ensure the marker is created even if parsing the args with NodeArgsParser fails.
+        if (optionSet.has(isRegistrationArg)) {
+            createNodeRegistrationMarker(baseDirectory)
+        }
+        return nodeRegistrationMarkerExists(baseDirectory)
+    }
+
+    private fun deleteNodeRegistrationMarker(baseDir: Path) {
+        try {
+            val marker = File((baseDir / INITIAL_REGISTRATION_MARKER).toUri())
```
Comment (kasiastreich) [inline node/src/main/kotlin/net/corda/node/internal/NodeStartup.kt]: Here and in `createNodeRegistrationMarker` you use `nodeRegistrationMarkerExist` logic, but you don't use that function.

File: node/src/main/kotlin/net/corda/node/internal/NodeStartup.kt [hunk 4]
```diff
@@ -184,6 +201,59 @@ open class NodeStartup(val args: Array<String>) {
...
+    private fun checkRegistrationMode(): Boolean {
+        // Parse the command line args just to get the base directory. The base directory is needed to determine
+        // if the node registration marker file exists, _before_ we call NodeArgsParser.parse().
+        // If it does exist, we call NodeArgsParser with `--initial-registration` added to the argument list. This way
+        // we make sure that the initial registration is completed, even if the node was restarted before the first
+        // attempt to register succeeded and the node administrator forgets to specify `--initial-registration` upon
+        // restart.
+        val optionParser = OptionParser()
+        optionParser.allowsUnrecognizedOptions()
+        val baseDirectoryArg = optionParser
```
Comment (kasiastreich) [inline node/src/main/kotlin/net/corda/node/internal/NodeStartup.kt]: I think but you can correct me, that we have two places where baseDirectory comes from. Primarly from config file (!) not command line flag.

File: node/src/main/kotlin/net/corda/node/internal/NodeStartup.kt [hunk 5]
```diff
@@ -65,13 +70,23 @@ open class NodeStartup(val args: Array<String>) {
             println("Corda will now exit...")
             return false
         }
-        val cmdlineOptions = NodeArgsParser().parseOrExit(*args)
+
+        val registrationMode = checkRegistrationMode()
+        val cmdlineOptions: CmdLineOptions
+        if (registrationMode && !args.contains("--initial-registration")) {
+            // The node was started before with `--initial-registration`, but the registration has not completed.
+            // So we pretend that the node was started with `--initial-registration`
```
Comment (kasiastreich) [inline node/src/main/kotlin/net/corda/node/internal/NodeStartup.kt]: If you forget initial-registration it adds it for you. However with initial registration you need network trustststore path and password, they kind of go together. Saving password is a no. Could you make sure that when node spits out massive error about option parsing failure, you print out instead some information what happened and how to deal with it?

File: node/src/main/kotlin/net/corda/node/internal/NodeStartup.kt (hunk 4 above)
Comment (florian-f) [reply node/src/main/kotlin/net/corda/node/internal/NodeStartup.kt]: `--baseDirectory` It has a default value (`.`), which is specified in `NodeArgsParser`. The same is used here. The docs don't mention `baseDirectory` as an entry in the config file, only that 

>  `baseDirectory` is available as a substitution value and contains the absolute path to the node’s base directory.

https://docs.corda.net/head/corda-configuration-file.html#fields

However, that does not mean that it's impossible to specify `baseDirectory` in the config file. I will investigate what happens in that case.

File: node/src/main/kotlin/net/corda/node/internal/NodeStartup.kt (hunk 1 above)
Comment (florian-f) [reply node/src/main/kotlin/net/corda/node/internal/NodeStartup.kt]: Good point, thanks.

File: node/src/main/kotlin/net/corda/node/internal/NodeStartup.kt (hunk 2 above)
Comment (florian-f) [reply node/src/main/kotlin/net/corda/node/internal/NodeStartup.kt]: Yes, you are right. Updating this.

File: node/src/main/kotlin/net/corda/node/internal/NodeStartup.kt (hunk 5 above)
Comment (florian-f) [reply node/src/main/kotlin/net/corda/node/internal/NodeStartup.kt]: Currently, it does not spit out a _massive_ error, but it complains that the password option is missing and prints the help message, just like when you would run `java -jar corda.jar --initial-registration`. There's not indication as to _why_ this is required though, but once I address your other comment (about printing the log statement to the console as well) there will be.

File: node/src/main/kotlin/net/corda/node/internal/NodeStartup.kt (hunk 3 above)
Comment (florian-f) [reply node/src/main/kotlin/net/corda/node/internal/NodeStartup.kt]: True, can be simplified.

File: node/src/main/kotlin/net/corda/node/internal/NodeStartup.kt (hunk 4 above)
Comment (florian-f) [reply node/src/main/kotlin/net/corda/node/internal/NodeStartup.kt]: OK I checked, putting `baseDirectory="/home/some/directory"` in the `node.conf` does not have any effect on `cmdlineOptions.baseDirectory`.

**Pull request merged**